app/lib/task/cloudcompute/googlecloudcompute.dart

```diff
@@ -0,0 +1,502 @@
+import 'package:retry/retry.dart';
+import 'package:ulid/ulid.dart';
+
+final _log = Logger('pub.googlecloudcompute');
+
+/// Hardcoded list of GCE zones to consider for provisioning.
+///
+/// See [list of regions and zones][regions-zones].
+///
+/// [regions-zones]: https://cloud.google.com/compute/docs/regions-zones
+const _googleCloudZones = [
+  'us-central1-a',
+  'us-central1-b',
+  'us-central1-c',
+  'us-central1-f',
+];
+
+/// Hardcoded machine type to use for provisioning instances.
+///
+/// Note. it is important that this _machine type_ is available in the
+/// [_googleCloudZones] listed.
+///
+/// See [predefined machine-types][machines-types].
+///
+/// [machine-types]: https://cloud.google.com/compute/docs/machine-types
+const _googleCloudMachineType = 'e2-standard-2'; // 2 vCPUs 8GB ram
+
+/// OAuth scope needed for the [http.Client] passed to
+/// [createGoogleCloudCompute].
+const googleCloudComputeScope = ComputeApi.cloudPlatformScope;
+
+/// Create a [CloudCompute] abstraction wrapping Google Compute Engine.
+///
+/// The [CloudCompute] abstraction created will manage instance in the given
+/// GCP cloud [project] with labels:
+///  * `owner = 'pub-dev'`, and,
+///  * `pool = poolLabel`.
+///
+/// Similarly, all instances created by this abstraction will be labelled with
+/// `owner` as `'pub-dev'` and `pool` as [poolLabel]. This allows for multiple
+/// pools of machines that don't interfere with eachother. By using a
+/// [poolLabel] such as `'<runtimeVersion>/pana'` we can ensure that the
+/// [CloudCompute] object for pana-tasks doesn't interfere with the other
+/// _runtime versions_ in production.
+///
+/// Instances will use [Container-Optimized OS][c-o-s] to the docker-image
+/// specified at instance creation.
+///
+/// [c-o-s]: https://cloud.google.com/container-optimized-os/docs
+CloudCompute createGoogleCloudCompute({
+  required http.Client client,
+  required String project,
+  required String poolLabel,
+}) {
+  if (poolLabel.isEmpty) {
+    throw ArgumentError.value(poolLabel, 'poolLabel', 'must not be empty');
+  }
+
+  return _GoogleCloudCompute(
+    ComputeApi(client),
+    project,
+    _googleCloudZones,
+    _googleCloudMachineType,
+    poolLabel,
+  );
+}
+
+class _GoogleCloudInstance extends CloudInstance {
+  /// GCP zone this instance exists inside.
+  @override
+  final String zone;
+
+  @override
+  final String name;
+
+  @override
+  final DateTime created;
+
+  @override
+  final InstanceState state;
+
+  _GoogleCloudInstance(
+    this.zone,
+    this.name,
+    this.created,
+    this.state,
+  );
+
+  @override
+  String toString() {
+    return 'GoogleCloudInstance($name, zone: $zone, created: $created, state: $state)';
+  }
+}
+
+class _PendingGoogleCloudInstance extends CloudInstance {
+  /// GCP zone this instance exists inside.
+  @override
+  final String zone;
+
+  @override
+  final String name;
+
+  @override
+  DateTime get created => clock.now();
```

Comment:
Not sure if this is relevant anywhere, but I'd expect the `created` to be `final`, e.g. `late final created = clock.now();`

Reply:
But while pending the instance haven't been created yet.. I guess we could also use `DateTime(0)` as a special place-hoilder... or something... Or some future time...

But maybe, I should document this behavior.

Reply:
I think you should either make `created` nullable (null when instance has not been created) or raise a `StateError` for pending instances.

Reply:
maybe... I don't think I want to handle the corner case, the `created` time is only suitable for GC'ing old instances.